ibis/formats/__init__.py

```diff
@@ -0,0 +1,124 @@
+from abc import ABC, abstractmethod
+from typing import Generic, TypeVar
+
+from ibis.expr.datatypes import DataType
+from ibis.expr.schema import Schema
+
+T = TypeVar('T', covariant=True)
+S = TypeVar('S', covariant=True)
+VS = TypeVar('VS', covariant=True)
+VC = TypeVar('VC', covariant=True)
+VT = TypeVar('VT', covariant=True)
```

Comment:
I'm not sure we need the covariance, unless we're considering `Format[A]` to be a subtype of `Format[B]` if `A` is a subclass of `B`.